src/connector/src/sink/iceberg/mod.rs

```diff
@@ -1942,45 +2057,213 @@ impl TwoPhaseCommitCoordinator for IcebergSinkCommitter {
             return Ok(());
         }
 
-        let mut write_results_bytes = deserialize_metadata(commit_metadata);
-
-        let snapshot_id_bytes = write_results_bytes.pop().unwrap();
-        let snapshot_id = i64::from_le_bytes(
-            snapshot_id_bytes
-                .try_into()
-                .map_err(|_| SinkError::Iceberg(anyhow!("Invalid snapshot id bytes")))?,
+        let parsed_commit_metadata = Self::parse_commit_metadata(commit_metadata)?;
+        tracing::info!(
+            "Commit metadata has_schema_change: {}, add_columns: {:?}",
+            parsed_commit_metadata.has_schema_change(),
+            parsed_commit_metadata
+                .add_columns
+                .as_ref()
+                .map(|cols| cols.iter().map(|c| &c.name).collect::<Vec<_>>())
         );
 
-        if self
+        if !parsed_commit_metadata.has_data() && !parsed_commit_metadata.has_schema_change() {
+            tracing::debug!(?epoch, "no data and no schema change to commit");
+            return Ok(());
+        }
+
+        let data_file_need_committed = parsed_commit_metadata.has_data();
+        let schema_change_need_commit = parsed_commit_metadata.has_schema_change();
+        let write_results = parsed_commit_metadata.write_results;
+        let snapshot_id = parsed_commit_metadata.snapshot_id;
+        let add_columns = parsed_commit_metadata.add_columns;
+
+        match (data_file_need_committed, schema_change_need_commit) {
+            (true, false) => {
+                // only commit data
+                let snapshot_id = snapshot_id
+                    .ok_or_else(|| SinkError::Iceberg(anyhow!("Snapshot id missing")))?;
+                self.commit_data_only(epoch, write_results, snapshot_id)
+                    .await
+            }
+            (false, true) => {
+                // only commit schema change
+                let add_columns = add_columns
+                    .ok_or_else(|| SinkError::Iceberg(anyhow!("Schema metadata missing")))?;
+                self.commit_schema_only(epoch, add_columns).await
+            }
+            (true, true) => {
+                // commit data and schema change
+                let snapshot_id = snapshot_id
+                    .ok_or_else(|| SinkError::Iceberg(anyhow!("Snapshot id missing")))?;
+                let add_columns = add_columns
+                    .ok_or_else(|| SinkError::Iceberg(anyhow!("Schema metadata missing")))?;
+                self.commit_data_and_schema(epoch, write_results, snapshot_id, add_columns)
+                    .await
+            }
+            (false, false) => Ok(()),
+        }
+    }
+
+    async fn abort(&mut self, _epoch: u64, _commit_metadata: Vec<u8>) {
+        // TODO: Files that have been written but not committed should be deleted.
+        tracing::debug!("Abort not implemented yet");
+    }
+}
+
+impl IcebergSinkCommitter {
+    fn parse_commit_metadata(commit_metadata: Vec<u8>) -> Result<ParsedCommitMetadata> {
```

Comment:
We need to better design the commit metadata serialization, because it would be persisted  in the metastore, so backward compatibility is needed. The current implementation breaks the compatibility which is undesired.